src/krb5_plugin/idp/idp_kdcpreauth.c

```diff
@@ -0,0 +1,812 @@
+                       void *data)
+{
+    static bool verify_success = true;
+    static bool verify_failure = false;
+    struct sss_idpkdc_verify *state;
+    krb5_kdcpreauth_modreq modreq;
+    krb5_error_code ret;
+    int i;
+
+    state = (struct sss_idpkdc_verify *)data;
+    modreq = (krb5_kdcpreauth_modreq)&verify_failure;
+
+    if (rret != 0) {
+        ret = rret;
+        goto done;
+    }
+
+    if (krad_packet_get_code(rres) != krad_code_name2num("Access-Accept")) {
+        ret = KRB5_PREAUTH_FAILED;
+        goto done;
+    }
+
+    state->enc_tkt_reply->flags |= TKT_FLG_PRE_AUTH;
+
+    for (i = 0; state->config->idpcfg->indicators[i] != NULL; i++) {
+        ret = state->cb->add_auth_indicator(state->kctx, state->rock,
+                                            state->config->idpcfg->indicators[i]);
+        if (ret != 0) {
+            goto done;
+        }
+    }
+
+    modreq = (krb5_kdcpreauth_modreq)&verify_success;
+    ret = 0;
+
+done:
+    state->respond(state->arg, ret, modreq, NULL, NULL);
+    sss_idpkdc_verify_free(state);
+    return;
+}
+
+/* Send Access-Request with password and state set to indicate that the user has
+ * finished authentication against idp provider. We expect Access-Accept. */
+static krb5_error_code
+sss_idpkdc_verify_send(krb5_context kctx,
+                       verto_ctx *vctx,
+                       krb5_kdcpreauth_rock rock,
+                       krb5_kdcpreauth_callbacks cb,
+                       krb5_enc_tkt_part *enc_tkt_reply,
+                       krb5_kdcpreauth_verify_respond_fn respond,
+                       void *arg,
+                       const krb5_data *rstate,
+                       struct sss_idpkdc_config *config)
+{
+    struct sss_idpkdc_verify *state;
+    krb5_error_code ret;
+
+    state = malloc(sizeof(struct sss_idpkdc_verify));
+    if (state == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+    memset(state, 0, sizeof(struct sss_idpkdc_verify));
+
+    state->config = config;
+    state->kctx = kctx;
+    state->rock = rock;
+    state->cb = cb;
+    state->enc_tkt_reply = enc_tkt_reply;
+    state->respond = respond;
+    state->arg = arg;
+
+    state->radius = sss_idpkdc_radius_init(kctx, vctx, config);
+    if (state->radius == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = krad_attrset_add(state->radius->attrs, krad_attr_name2num("State"),
+                           rstate);
+    if (ret != 0) {
+        goto done;
+    }
+
+    ret = krad_client_send(state->radius->client,
+                           krad_code_name2num("Access-Request"),
+                           state->radius->attrs, config->server,
+                           config->secret, config->timeout, config->retries,
+                           sss_idpkdc_verify_done, state);
+
+done:
+    if (ret != 0) {
+        /* It is the caller responsibility to free config in case of error. */
+        state->config = NULL;
```

Comment:
Hi,

you have to check if `state != NULL` before or immediately return if the initial malloc() failed.

bye,
Sumit